tests/server/routes/post-metrics.js

```diff
@@ -198,29 +233,596 @@ define([
+          },
+
+          'metricsCollector.write was called': () => {
+            assert.strictEqual(mocks.metricsCollector.write.callCount, 1);
+          },
+
+          'statsdCollector.write was called': () => {
+            assert.strictEqual(mocks.statsdCollector.write.callCount, 1);
+          },
+
+          'gaCollector.write was called': () => {
+            assert.strictEqual(mocks.gaCollector.write.callCount, 1);
+          },
+
+          'flowEvent was not called': () => {
+            assert.strictEqual(mocks.flowEvent.callCount, 0);
+          }
+        }
+      },
+
+      'route.process without optional flow data': {
+        setup: function () {
+          sinon.stub(Date, 'now', function () {
+            return 1000;
+          });
+          setupMetricsHandlerTests({
+            /*eslint-disable sorting/sort-object-props*/
+            data: {
+              flowBeginTime: 42,
+              flowId: '1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef',
+              isSampledUser: true,
+              startTime: 10,
+              flushTime: 20
+            },
+            events: [
+              { type: 'foo', offset: 0 },
+              { type: 'bar', offset: 1 },
+              { type: 'flow.signup.begin', offset: 2 },
+              { type: 'baz', offset: 3 },
+              { type: 'flow.signup.engage', offset: 11 },
+              { type: 'flow.signup.submit', offset: 4 }
+            ],
+            userAgent: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:47.0) Gecko/20100101 Firefox/47.0'
+            /*eslint-enable sorting/sort-object-props*/
+          });
+        },
+
+        teardown () {
+          Date.now.restore();
+          sandbox.reset();
+        },
+
+        'response.json was called': function () {
+          assert.equal(mocks.response.json.callCount, 1);
+        },
+
+        'process.nextTick was called': function () {
+          assert.equal(mocks.nextTick.callCount, 1);
+        },
+
+        'process.nextTick callback': {
+          setup: function () {
+            flowMetricsCheckResult = true;
+            mocks.nextTick.args[0][0]();
+          },
+
+          'mozlog.error was not called': function () {
+            assert.strictEqual(mocks.mozlog.error.callCount, 0);
+          },
+
+          'metricsCollector.write was called': function () {
+            assert.strictEqual(mocks.metricsCollector.write.callCount, 1);
+          },
+
+          'statsdCollector.write was called': function () {
+            assert.strictEqual(mocks.statsdCollector.write.callCount, 1);
+          },
+
+          'gaCollector.write was called': function () {
+            assert.strictEqual(mocks.gaCollector.write.callCount, 1);
+          },
+
+          'flowEvent was called': function () {
+            assert.strictEqual(mocks.flowEvent.callCount, 2);
+          }
+        }
+      },
+
+      'route.process with valid-seeming flow data but flowMetrics.check returns false': {
```

Comment:
There is a common set of assertions in the test cases and most have a similar layout. I have used dynamic tests in similar situations to help reduce duplication. That being said, not sure if that is possible here or if you are opposed to it.